cli/src/pcluster/cli/entrypoint.py

```diff
@@ -1,155 +1,236 @@
+        file_data = file.read()
+    return base64.b64encode(file_data.encode('utf-8')).decode('utf-8')
+
+
+def convert_args(model, op_name, args_in):
+    """Takes a model, the name of the operation and the arguments
+    provided by argparse and converts the parameters into a format
+    that is suitable to be called in the controllers."""
+    body = {}
+    kwargs = {}
+    for param in model[op_name]['params']:
+        param_name = to_snake_case(param['name'])
+        value = args_in.pop(param_name)
+
+        if param['body']:
+            param_name = camelcase(param_name)
+            param_name = param_name[0].lower() + param_name[1:]
+            body[param_name] = value
+        else:
+            kwargs[param_name] = value
+
+    kwargs.update(args_in)
+
+    return body, kwargs
+
+
+def dispatch(model, args):
+    """Dispatches to a controller function when the arguments have an
+    operation specified."""
+    args_dict = args.__dict__
+    operation = args.operation
+    del args_dict['func']
+    del args_dict['operation']
+    body, kwargs = convert_args(model, operation, args_dict)
+
+    dispatch_func = partial(pcluster.cli.model.call, model[operation]['func'])
+    if len(body) > 0:
+        dispatch_func = partial(dispatch_func, body)
+
+    # middleware provides an opportunity to customize the calling of the
+    # underlying API function on a per-operation basis
+    middleware_funcs = inspect.getmembers(pcluster.cli.middleware,
+                                          inspect.isfunction)
+    middleware = {to_kebab_case(f[0]): f[1] for f in middleware_funcs}
+
+    if operation in middleware:
+        return middleware[operation](dispatch_func, body, kwargs)
+    else:
+        return dispatch_func(**kwargs)
+
+
+def gen_parser(model):
+    """Takes a model and returns an ArgumentParser for converting command line
+    arguments into values based on that model."""
+    desc = ("pcluster is the AWS ParallelCluster CLI and permits "
+            "launching and management of HPC clusters in the AWS cloud.")
+    epilog = 'For command specific flags, please run: "pcluster [command] --help"'
+    parser = argparse.ArgumentParser(description=desc, epilog=epilog)
+    subparsers = parser.add_subparsers(help="", required=True, title='COMMANDS',
+                                       dest='operation')
+    type_map = {'int': int, 'boolean': bool_converter, 'byte': read_file_b64}
+    parser_map = {'subparser': subparsers}
+
+    # Add each operation as it's onn parser with params / body as arguments
+    for op_name, operation in model.items():
+        op_help = operation.get('description', f"{op_name} command help")
+        subparser = subparsers.add_parser(op_name, help=op_help, description=op_help)
+        parser_map[op_name] = subparser
+
+        for param in operation['params']:
+            help = param.get('description', '')
+            metavar = param['name'].upper() if len(param.get('enum', [])) > 4 else None
```

Comment:
Some of the arguments (like the region options) were getting to be pretty long and made the help output cumbersome so it would say replace the list of regions with `REGION`. We certainly don't need to keep it and I can see why we might want to see all of them even if it makes the output ugly.